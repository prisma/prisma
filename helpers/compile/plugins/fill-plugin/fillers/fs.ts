export function existsSync() {
  return false
}

/**
 * A stub for fs for tryLoadEnv not to attempt loading .env
 */
const fs = {
  existsSync() {
    return false
  },
}

export default fs
