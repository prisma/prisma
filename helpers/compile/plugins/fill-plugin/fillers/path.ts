export function resolve(...args: any[]) {
  return args.join('/')
}

export function join(...args: any[]) {
  return args.join('/')
}

export const sep = '/'

export const posix = {
  sep,
}

/**
 * A poor man's shim for the "path" module
 */
const path = {
  resolve,
  posix,
  join,
  sep,
}

export default path
