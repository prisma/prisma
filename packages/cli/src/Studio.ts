import Debug from '@prisma/debug'
import { enginesVersion } from '@prisma/engines'
import {
  arg,
  Command,
  format,
  getConfig,
  getDirectUrl,
  HelpError,
  isError,
  loadEnvFile,
  resolveUrl,
} from '@prisma/internals'
import { getSchemaPathAndPrint } from '@prisma/migrate'
import { StudioServer } from '@prisma/studio-server'
import fs from 'fs'
import getPort from 'get-port'
import { bold, dim, red } from 'kleur/colors'
import open from 'open'
import path from 'path'

// Note that we have a test relying on the namespace
// Any change to the namespace must be done in the test as well
// See packages/client/tests/e2e/issues/studio-1128-spawn-enoent/_steps.ts
const debug = Debug('prisma:cli:studio')

const packageJson = require('../package.json') // eslint-disable-line @typescript-eslint/no-var-requires

export class Studio implements Command {
  public instance?: StudioServer

  public static new(): Studio {
    return new Studio()
  }

  private static help = format(`
Browse your data with Prisma Studio

${bold('Usage')}

  ${dim('$')} prisma studio [options]

${bold('Options')}

  -h, --help        Display this help message
  -p, --port        Port to start Studio on
  -b, --browser     Browser to open Studio in
  -n, --hostname    Hostname to bind the Express server to
  --schema          Custom path to your Prisma schema

${bold('Examples')}

  Start Studio on the default port
    ${dim('$')} prisma studio

  Start Studio on a custom port
    ${dim('$')} prisma studio --port 5555

  Start Studio in a specific browser
    ${dim('$')} prisma studio --port 5555 --browser firefox
    ${dim('$')} BROWSER=firefox prisma studio --port 5555

  Start Studio without opening in a browser
    ${dim('$')} prisma studio --port 5555 --browser none
    ${dim('$')} BROWSER=none prisma studio --port 5555

  Specify a schema
    ${dim('$')} prisma studio --schema=./schema.prisma
`)

  /**
   * Parses arguments passed to this command, and starts Studio
   *
   * @param argv Array of all arguments
   */
  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--port': Number,
      '-p': '--port',
      '--browser': String,
      '-b': '--browser',
      '--hostname': String,
      '-n': '--hostname',
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    loadEnvFile({ schemaPath: args['--schema'], printMessage: true })

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const hostname = args['--hostname']
    const port = args['--port'] || (await getPort({ port: getPort.makeRange(5555, 5600) }))
    const browser = args['--browser'] || process.env.BROWSER

    const staticAssetDir = path.resolve(__dirname, '../build/public')

    const schema = await fs.promises.readFile(schemaPath, 'utf-8')
    const config = await getConfig({ datamodel: schema, ignoreEnvVarErrors: true })

    process.env.PRISMA_DISABLE_WARNINGS = 'true' // disable client warnings
    const studio = new StudioServer({
      schemaPath,
      hostname,
      port,
      staticAssetDir,
      prismaClient: {
        resolve: {
          '@prisma/client': path.resolve(__dirname, '../prisma-client/index.js'),
        },
        directUrl: resolveUrl(getDirectUrl(config.datasources[0])),
      },
      versions: {
        prisma: packageJson.version,
        queryEngine: enginesVersion,
      },
    })

    await studio.start()

    const serverUrl = `http://localhost:${port}`
    if (!browser || browser.toLowerCase() !== 'none') {
      try {
        const subprocess = await open(serverUrl, {
          app: browser,
          url: true,
        })

        subprocess.on('spawn', () => {
          // We match on this string in debug logs in tests
          debug(`requested to open the url ${serverUrl}`)
        })

        subprocess.on('error', (e) => {
          debug(e)
          // We match on this string in debug logs in tests
          debug(`failed to open the url ${serverUrl} in browser`)
        })
      } catch (e) {
        // Ignore any errors that occur when trying to open the browser, since they should not halt the process
        debug(e)
      }
    }

    this.instance = studio

    return `Prisma Studio is up on ${serverUrl}`
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${bold(red(`!`))} ${error}\n${Studio.help}`)
    }

    return Studio.help
  }
}
