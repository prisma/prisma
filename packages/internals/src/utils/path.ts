import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
