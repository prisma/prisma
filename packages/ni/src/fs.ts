import { resolve } from 'node:path'
import fs from 'node:fs'
import type { RunnerContext } from './runner'

export function getPackageJSON(ctx?: RunnerContext): any {
  const cwd = ctx?.cwd ?? process.cwd()
  const path = resolve(cwd, 'package.json')

  if (fs.existsSync(path)) {
    try {
      const raw = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(raw)
      return data
    }
    catch (e) {
      if (!ctx?.programmatic) {
        console.warn('Failed to parse package.json')
        process.exit(1)
      }

      throw e
    }
  }
}
