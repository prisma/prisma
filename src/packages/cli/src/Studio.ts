import { getPlatform } from '@prisma/get-platform'
import {
  arg,
  Command,
  format,
  getSchemaPath,
  HelpError,
  isError,
  ProviderAliases,
} from '@prisma/sdk'
import StudioServer from '@prisma/studio-server'
import chalk from 'chalk'
import getPort from 'get-port'
import open from 'open'
import path from 'path'

const packageJson = require('../package.json') // eslint-disable-line @typescript-eslint/no-var-requires

export class Studio implements Command {
  public instance?: StudioServer

  public static new(providerAliases: ProviderAliases): Studio {
    return new Studio(providerAliases)
  }

  private static help = format(`
    Browse your data with Prisma Studio

    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma studio

    ${chalk.bold('Options')}

      -h, --help        Displays this help message
      -p, --port        Port to start Studio on
      -b, --browser     Browser to open Studio in

    ${chalk.bold('Examples')}

      Start Studio on the default port
      ${chalk.dim('$')} prisma studio

      Start Studio on a custom port
      ${chalk.dim('$')} prisma studio --port 5555

      Start Studio in a specific browser
      ${chalk.dim('$')} prisma studio --port 5555 --browser firefox
      ${chalk.dim('$')} BROWSER=firefox prisma studio --port 5555

      Start Studio without opening in a browser
      ${chalk.dim('$')} prisma studio --port 5555 --browser none
      ${chalk.dim('$')} BROWSER=none prisma studio --port 5555
  `)

  private constructor(private readonly providerAliases: ProviderAliases) {
    this.providerAliases = providerAliases
  }

  /**
   * Parses arguments passed to this command, and starts Studio
   *
   * @param argv Array of all arguments
   */
  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--port': Number,
      '-p': '--port',
      '--browser': String,
      '-b': '--browser',
      '--schema': String,
      '--experimental': Boolean,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (args['--experimental']) {
      console.warn(
        `${chalk.yellow(
          'warn',
        )} --experimental is no longer required for this command as Studio is now Generally Available.`,
      )
    }

    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      throw new Error(
        `Could not find a ${chalk.bold(
          'schema.prisma',
        )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
          '--schema',
        )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
          './prisma/schema.prisma',
        )} https://pris.ly/d/prisma-schema-location`,
      )
    }

    console.log(
      chalk.dim(
        `Prisma schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    const port =
      args['--port'] || (await getPort({ port: getPort.makeRange(5555, 5600) }))
    const browser = args['--browser'] || process.env.BROWSER
    const platform = await getPlatform()
    const extension = platform === 'windows' ? '.exe' : ''
    const queryEnginePath =
      process.env.NODE_ENV === 'production'
        ? eval(
            `require('path').join(__dirname, '../query-engine-${platform}${extension}')`,
          )
        : eval(
            `require('path').join(__dirname, '../node_modules/@prisma/sdk/query-engine-${platform}${extension}')`,
          )
    const staticAssetDir =
      process.env.NODE_ENV === 'production'
        ? path.resolve(__dirname, './public')
        : path.resolve(__dirname, '../dist/public')

    const studio = new StudioServer({
      schemaPath,
      port,
      prismaClient: {
        generator: {
          version: packageJson.prisma.version,
          providerAliases: this.providerAliases,
        },
      },
      binaryPaths: {
        queryEngine: queryEnginePath,
      },
      staticAssetDir,
      versions: {
        prisma2: packageJson.version,
        queryEngine: packageJson.prisma.version,
      },
    })

    await studio.start()

    const serverUrl = `http://localhost:${port}`
    if (!browser || browser.toLowerCase() !== 'none') {
      await open(serverUrl, {
        app: browser,
        url: true,
        // eslint-disable-next-line @typescript-eslint/no-empty-function
      }).catch(() => {}) // Ignore any errors
    }

    this.instance = studio

    return `Prisma Studio is up on ${serverUrl}`
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${Studio.help}`)
    }

    return Studio.help
  }
}
