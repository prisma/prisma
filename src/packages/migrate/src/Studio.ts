import Debug from '@prisma/debug'
import fs from 'fs'
import getPort from 'get-port'
import path from 'path'
import open from 'open'

import { getSchemaPathSync } from '@prisma/sdk'
import { getPlatform } from '@prisma/get-platform'
import { ProviderAliases } from '@prisma/sdk'

export interface StudioOptions {
  schemaPath?: string
  browser?: string
  port?: number
}

const debug = Debug('Studio')
const packageJson = eval(`require('../package.json')`) // tslint:disable-line

export class Studio {
  private schemaPath: string
  private browser?: string
  private instance?: any
  private port?: number

  constructor({ schemaPath, browser, port }: StudioOptions = {}) {
    this.schemaPath = this.getSchemaPath(schemaPath)
    this.browser = browser
    this.port = port
  }

  public async start(providerAliases: ProviderAliases): Promise<string> {
    try {
      if (this.instance) {
        throw new Error(`Studio is already started`)
      }

      const platform = await getPlatform()
      const extension = platform === 'windows' ? '.exe' : ''

      const pathCandidates = [
        // ncc go home
        // tslint:disable-next-line
        eval(
          `require('path').join(__dirname, '../node_modules/@prisma/sdk/query-engine-${platform}${extension}')`,
        ), // for local dev
        // tslint:disable-next-line
        eval(
          `require('path').join(__dirname, '../query-engine-${platform}${extension}')`,
        ), // for production
      ]

      const pathsExist = await Promise.all(
        // eslint-disable-next-line @typescript-eslint/require-await
        pathCandidates.map(async (candidate) => ({
          exists: fs.existsSync(candidate),
          path: candidate,
        })),
      )

      const firstExistingPath = pathsExist.find((p) => p.exists)

      if (!firstExistingPath) {
        throw new Error(
          `Could not find any Prisma query-engine binary for Studio. Looked in ${pathCandidates.join(
            ', ',
          )}`,
        )
      }

      const StudioServer = (await import('@prisma/studio-server')).default

      if (!this.port) {
        this.port = await getPort({ port: getPort.makeRange(5555, 5600) })
      }

      this.instance = new StudioServer({
        port: this.port,
        schemaPath: this.schemaPath,
        binaryPaths: {
          queryEngine: firstExistingPath.path,
        },
        prismaClient: {
          generator: {
            version: packageJson.prisma.version,
            providerAliases,
          },
        },
        staticAssetDir: path.resolve(__dirname, 'public'), // Overriding this directory since after NCC compilation, this won't resolve automatically
        versions: {
          prisma2: packageJson.version,
          queryEngine: packageJson.prisma.version,
        },
      })

      await this.instance.start()

      const serverUrl = `http://localhost:${this.port}`
      switch (this.browser) {
        case 'none':
        case 'NONE':
          break
        default:
          await open(serverUrl, {
            app: this.browser,
            url: true,
            // eslint-disable-next-line @typescript-eslint/no-empty-function
          }).catch(() => {}) // Ignore any errors
          break
      }

      return `Studio started at ${serverUrl}`
    } catch (e) {
      debug(e)
    }

    return ''
  }

  public async restart(providerAliases: ProviderAliases): Promise<string> {
    if (this.instance) {
      await this.instance.stop('Restarting')
      await this.instance.start(providerAliases)
      return ''
    }

    return this.start(providerAliases)
  }

  private getSchemaPath(schemaPathFromOptions?): string {
    const schemaPath = getSchemaPathSync(schemaPathFromOptions)
    if (!schemaPath) {
      throw new Error(
        `Could not find ${schemaPathFromOptions || 'schema.prisma'}`,
      )
    }

    return schemaPath
  }
}
