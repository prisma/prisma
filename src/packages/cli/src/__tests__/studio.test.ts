import fs from 'fs'
import http from 'http'
import path from 'path'
import WebSocket from 'ws'
import rimraf from 'rimraf'

import { Studio } from '../Studio'

const STUDIO_TEST_PORT = 5678

// silencium
console.log = () => null

const setupWS = (): Promise<WebSocket> => {
  return new Promise(res => {
    const ws = new WebSocket(`ws://127.0.0.1:${STUDIO_TEST_PORT}/`)
    ws.on('open', () => {
      ws.on('message', (data: string) => {
        const message: any = JSON.parse(data)

        expect(message.channel).not.toBeUndefined()
        expect(message.action).not.toBeUndefined()

        if (message.channel !== '-photon' && message.action !== 'start') {
          return
        }

        res(ws)
      })

      ws.send(
        JSON.stringify({
          requestId: 1,
          channel: 'photon',
          action: 'start',
          payload: {},
        }),
      )
    })
  })
}

const sendRequest = (ws: WebSocket, message: any): Promise<any> => {
  return new Promise(res => {
    ws.on('message', (data: string) => {
      const message: any = JSON.parse(data)

      expect(message.channel).not.toBeUndefined()
      expect(message.action).not.toBeUndefined()

      if (message.channel !== '-photon' && message.action !== 'request') {
        return
      }

      res(message)
    })

    // Send the same query Studio would send if launched
    ws.send(JSON.stringify(message))
  })
}

let studio: Studio
let ws: WebSocket

beforeEach(async () => {
  // Before  every test, we'd like to reset the DB.
  // We do this by duplicating the original SQLite DB file, and using the duplicate as the datasource in our schema
  rimraf.sync(path.join(__dirname, './fixtures/studio-test-project/dev_tmp.db'))
  fs.copyFileSync(
    path.join(__dirname, './fixtures/studio-test-project/dev.db'),
    path.join(__dirname, './fixtures/studio-test-project/dev_tmp.db'),
  )

  // Clean up Client generation directory
  rimraf.sync(path.join(__dirname, '../prisma-client'))
  studio = Studio.new({
    // providerAliases
    'prisma-client-js': {
      generatorPath: `node --max-old-space-size=8096 "${path.join(
        __dirname,
        '../../../client/generator-build/index.js',
      )}"`,
      outputPath: path.join(__dirname, '../prisma-client/'),
    },
  })

  await studio.parse([
    '--schema',
    path.join(__dirname, './fixtures/studio-test-project/schema.prisma'),
    '--port',
    `${STUDIO_TEST_PORT}`,
    '--browser',
    'none',
  ])

  ws = await setupWS()
})

afterEach(async () => {
  await studio.instance?.stop()
  ws?.close()
})

it('launches client correctly', async () => {
  await new Promise<void>((res, rej) => {
    http.get(`http://localhost:${STUDIO_TEST_PORT}`, response => {
      try {
        expect(response.statusCode).toEqual(200)
        res()
      } catch (e) {
        rej(e)
      }
    })
  })
})

it('can respond to `findMany` queries', async () => {
  const res = await sendRequest(ws, {
    requestId: 1,
    channel: 'photon',
    action: 'request',
    payload: {
      data: {
        query: `
          prisma.with_all_field_types.findMany({
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            }
          })`,
      },
    },
  })
  expect(res).toMatchSnapshot()
})

it('can respond to `create` queries', async () => {
  const res = await sendRequest(ws, {
    requestId: 1,
    channel: 'photon',
    action: 'request',
    payload: {
      data: {
        query: `
          prisma.with_all_field_types.create({
            data: {
              id: 3,
              string: "",
              int: 0,
              float: 0.0,
              datetime: "2020-08-03T00:00:00.000Z",
              relation: {
                connect: {
                  id: 3
                }
              },
              relation_list: {
                connect: {
                  id: 3
                }
              }
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            }
          })`,
      },
    },
  })
  expect(res).toMatchSnapshot()
})

it('can respond to `update` queries', async () => {
  const res = await sendRequest(ws, {
    requestId: 1,
    channel: 'photon',
    action: 'request',
    payload: {
      data: {
        query: `
          prisma.with_all_field_types.update({
            where: {
              id: 1
            },
            data: {
              string: "Changed String",
              int: 100,
              float: 100.5,
              datetime: "2025-08-03T00:00:00.000Z",
              relation: {
                connect: {
                  id: 3
                }
              },
              relation_list: {
                connect: {
                  id: 3
                }
              }
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            }
          })`,
      },
    },
  })
  expect(res).toMatchSnapshot()
})

it('can respond to `delete` queries', async () => {
  const res = await sendRequest(ws, {
    requestId: 1,
    channel: 'photon',
    action: 'request',
    payload: {
      data: {
        query: `
          prisma.with_all_field_types.delete({
            where: { id: 2 },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            }
          })`,
      },
    },
  })
  expect(res).toMatchSnapshot()
})
