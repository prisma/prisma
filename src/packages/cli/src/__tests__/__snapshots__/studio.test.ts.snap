// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can respond to \`create\` queries 1`] = `
Object {
  action: request,
  channel: -photon,
  payload: Object {
    data: Object {
      params: Object {
        action: create,
        args: Object {
          data: Object {
            datetime: 2020-08-03T00:00:00.000Z,
            float: 0,
            id: 3,
            int: 0,
            relation: Object {
              connect: Object {
                id: 3,
              },
            },
            relation_list: Object {
              connect: Object {
                id: 3,
              },
            },
            string: ,
          },
          select: Object {
            datetime: true,
            float: true,
            id: true,
            int: true,
            relation: true,
            relation_list: true,
            string: true,
          },
        },
        clientMethod: with_all_field_types.create,
        dataPath: Array [],
        model: with_all_field_types,
        runInTransaction: false,
      },
      response: Object {
        datetime: 2020-08-03T00:00:00.000Z,
        float: 0,
        id: 3,
        int: 0,
        relation: Object {
          id: 3,
          name: Relation Target 003,
          waft_id: 3,
        },
        relation_list: Array [
          Object {
            id: 3,
            name: Relation Target 003,
            waft_id: 3,
          },
        ],
        string: ,
      },
    },
    error: null,
  },
  requestId: 1,
}
`;

exports[`can respond to \`delete\` queries 1`] = `
Object {
  action: request,
  channel: -photon,
  payload: Object {
    data: Object {
      params: Object {
        action: delete,
        args: Object {
          select: Object {
            datetime: true,
            float: true,
            id: true,
            int: true,
            relation: true,
            relation_list: true,
            string: true,
          },
          where: Object {
            id: 2,
          },
        },
        clientMethod: with_all_field_types.delete,
        dataPath: Array [],
        model: with_all_field_types,
        runInTransaction: false,
      },
      response: Object {
        datetime: 1970-01-01T00:00:00.000Z,
        float: 0,
        id: 2,
        int: 0,
        relation: Object {
          id: 2,
          name: Relation Target 002,
          waft_id: 2,
        },
        relation_list: Array [],
        string: Delete me,
      },
    },
    error: null,
  },
  requestId: 1,
}
`;

exports[`can respond to \`findMany\` queries 1`] = `
Object {
  action: request,
  channel: -photon,
  payload: Object {
    data: Object {
      params: Object {
        action: findMany,
        args: Object {
          select: Object {
            datetime: true,
            float: true,
            id: true,
            int: true,
            relation: true,
            relation_list: true,
            string: true,
          },
        },
        clientMethod: with_all_field_types.findMany,
        dataPath: Array [],
        model: with_all_field_types,
        runInTransaction: false,
      },
      response: Array [
        Object {
          datetime: 2020-08-03T00:00:00.000Z,
          float: 3.14,
          id: 1,
          int: 42,
          relation: Object {
            id: 1,
            name: Relation Target 001,
            waft_id: 1,
          },
          relation_list: Array [],
          string: Some string,
        },
        Object {
          datetime: 1970-01-01T00:00:00.000Z,
          float: 0,
          id: 2,
          int: 0,
          relation: Object {
            id: 2,
            name: Relation Target 002,
            waft_id: 2,
          },
          relation_list: Array [],
          string: Delete me,
        },
      ],
    },
    error: null,
  },
  requestId: 1,
}
`;

exports[`can respond to \`update\` queries 1`] = `
Object {
  action: request,
  channel: -photon,
  payload: Object {
    data: Object {
      params: Object {
        action: update,
        args: Object {
          data: Object {
            datetime: 2025-08-03T00:00:00.000Z,
            float: 100.5,
            int: 100,
            relation: Object {
              connect: Object {
                id: 3,
              },
            },
            relation_list: Object {
              connect: Object {
                id: 3,
              },
            },
            string: Changed String,
          },
          select: Object {
            datetime: true,
            float: true,
            id: true,
            int: true,
            relation: true,
            relation_list: true,
            string: true,
          },
          where: Object {
            id: 1,
          },
        },
        clientMethod: with_all_field_types.update,
        dataPath: Array [],
        model: with_all_field_types,
        runInTransaction: false,
      },
      response: Object {
        datetime: 2025-08-03T00:00:00.000Z,
        float: 100.5,
        id: 1,
        int: 100,
        relation: Object {
          id: 3,
          name: Relation Target 003,
          waft_id: 1,
        },
        relation_list: Array [
          Object {
            id: 3,
            name: Relation Target 003,
            waft_id: 1,
          },
        ],
        string: Changed String,
      },
    },
    error: null,
  },
  requestId: 1,
}
`;
